### PR TITLE
Fix the stackoverflow problem

### PR DESCRIPTION
**Pull request by michaelosthege**
See the commit messages for details.

### CODE REVIEW

File: pymc3/distributions/posterior_predictive.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
 )
 from ..exceptions import IncorrectArgumentsError
 from ..vartypes import theano_constant
-from ..util import dataset_to_point_dict, chains_and_samples, get_var_name
+from ..util import dataset_to_point_list, chains_and_samples, get_var_name
```
Comment (MarcoGorelli) [inline pymc3/distributions/posterior_predictive.py:45]: Is this change related to the windows bug? I'm still fine with it being here, it's just for my understanding :smile:

Comment (michaelosthege) [reply pymc3/distributions/posterior_predictive.py:45]: no, unrelated.. I just noticed the `FIXME` comment and wanted to get a DeprecationWarning in before 3.10
(without stressing the CI too much)

File: pymc3/tests/test_memo.py [hunk 2]
```diff
@@ -11,21 +11,59 @@
...
+    with pm.Model() as pmodel:
+        mu = pm.Normal("mu", 0, 1)
+        sd = pm.Gamma("sd", 1, 2)
+        dd = pm.DensityDist(
+            "dd",
+            pm.Normal.dist(mu, sd).logp,
+            random=pm.Normal.dist(mu, sd).random,
+            observed=obs,
+        )
+        print()
```
Comment (MarcoGorelli) [inline pymc3/tests/test_memo.py:60]: do these print statements have any effect in this test?

Comment (michaelosthege) [reply pymc3/tests/test_memo.py:60]: I added this print because the first print ended up on the same line when running `pytest -v -s`. It was hard to read..

Comment (michaelosthege) [reply pymc3/tests/test_memo.py:60]: And I printed because otherwise one doesn't know which variable caused the stackoverflow

File: pymc3/tests/test_memo.py [hunk 3]
```diff
@@ -11,21 +11,59 @@
...
+    assert len(fun.cache) == 1
+    assert funmem(inputs, "_b") == "['i1', 'i2']_b"
+    assert funmem(inputs, "_b") == "['i1', 'i2']_b"
+    assert len(fun.cache) == 2
+
+    # add items to the inputs list (the list instance remains identical !!)
+    inputs.append("i3")
+    assert funmem(inputs) == "['i1', 'i2', 'i3']_a"
+    assert funmem(inputs) == "['i1', 'i2', 'i3']_a"
+    assert len(fun.cache) == 3
```
Comment (MarcoGorelli) [inline pymc3/tests/test_memo.py:46]: really nice test here!
